target/linux/ipq807x/base-files/lib/upgrade/qnap.sh

```diff
@@ -0,0 +1,81 @@
+#
+# Copyright (C) 2016 lede-project.org
+#
+
+# qnap 301w has 3 kernel and 3 corresponding rootfs partitions. 
+# u-boot always provide the PARTUUID of the first rootfs parition as cmdline.
+# due to that we can not reliable use sysupgrade with A/B partitioning.
+# So lets use only the first kernel and rootfs partition 
+# 
+# this is based on the ipq806x zyxel.sh mmc upgrade
+
+qnap_do_upgrade() {
+	local tar_file="$1"
+	local board=$(board_name)
+	local rootfs=
+	local kernel=
+
+	case "$board" in
+	qnap,301w)
+			kernel="/dev/mmcblk0p1"
+			rootfs="/dev/mmcblk0p4"
```

Comment:
Can these be looked up by name instead of hardcoding?

Reply:
I could use the lately committed "find_mmc_part" function, but that would need a rebase of your branch on openwrt master.
Or I can implement something similar in the qnap.sh and can switch later to the find_mmc_part function if it's in your branch.

Reply:
I added the "find_mmc_part" function from openwrt master to the qnap.sh and switched to partition name.
If the generic "find_mmc_part" gets merged to your branch in the future, I'll remove the "find_mmc_part" from the qnap.sh.

Reply:
I have rebased the branch, so please use the upstream function and remove the hardcoded path as OpenWrt has an issue with those and if anything changes in the block device numbering then it simply breaks updating.

Reply:
done. rebased, using upstream function and removed hardcoded path (using part name instead)

Reply:
Is this script really special to QNAP or is it mostly copy/paste?

Reply:
It's mostly c&p from the zyxel.sh from the ipq806x target.

I think this could be generic, and we only need to pass kernel part name and rootfs name to the script.

And the loopdev offset check here:

```
local offset=$(tar xf $tar_file ${board_dir}/root -O | wc -c)
	[ $offset -lt **131072** ] 
```
The bold value needs to align to the blocksize in the image definition (otherwise openwrt doesn't recognize the ext4 overlay loopdev). But that's all. So we could maybe define that 3 as variables in the platform.sh and pass that to the script.

Reply:
That would be great to avoid duplicating the code

Reply:
Hi Robert,

I've pushed a few changes.

1. a mmc.sh as a generic mmc upgrade script
2. in platform.sh I've added a variable "kernelname" and "rootfsname" they will later be used in in the function "mmc_do_upgrade" (mmc.sh)
3. this means, the mmc_do_uggrade may also used for other devices, just define kernelname and rootfsname in platform.sh and call mmc_do_uggrade.

I hopes this is fine.

> > I am fine with having QNAP specific update script.
> 
> And then have basically the same script for netgear aswell, just with a different partition name in case that's needed? Sounds a little unusual to me, just wanted to double-check if I understood that correctly.

Hi, now we have generic mmc.sh, I hoped that is what you're looking for ;-)